File: src/main/java/com/dude/dms/backend/service/ChangelogService.java
package com.dude.dms.backend.service;

import com.dude.dms.backend.brain.DmsLogger;
import com.dude.dms.backend.data.Changelog;
import com.dude.dms.backend.repositories.ChangelogRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class ChangelogService extends CrudService<Changelog> {

    private static final DmsLogger LOGGER = DmsLogger.getLogger(ChangelogService.class);

    private final ChangelogRepository changelogRepository;

    @Autowired
    public ChangelogService(ChangelogRepository changelogRepository) {
        this.changelogRepository = changelogRepository;
    }

    @Override
    public JpaRepository<Changelog, Long> getRepository() {
        return changelogRepository;
    }

    public Optional<Changelog> findByVersion(String version) {
        return changelogRepository.findByVersion(version);
    }

    @Override
    public Changelog create(Changelog entity) {
        if (findByVersion(entity.getVersion()).isPresent()) {
            LOGGER.warn("Tried to create changelog with same Version");
            return null;
        }
        return super.create(entity);
    }
}

File: src/main/java/com/dude/dms/backend/service/ChangelogService.kt
package com.dude.dms.backend.service

import com.dude.dms.brain.DmsLogger.Companion.getLogger
import com.dude.dms.backend.data.Changelog
import com.dude.dms.backend.repositories.ChangelogRepository
import com.dude.dms.brain.events.EventManager
import org.springframework.stereotype.Service

@Service
class ChangelogService(
        private val changelogRepository: ChangelogRepository,
        eventManager: EventManager
) : EventService<Changelog>(changelogRepository, eventManager) {

    fun findByVersion(version: String) = changelogRepository.findByVersion(version)

    override fun create(entity: Changelog): Changelog {
        val current = findByVersion(entity.version)
        return if (current != null) {
            LOGGER.warn("Tried to create changelog with same Version")
            current
        } else super.create(entity)
    }

    companion object {
        private val LOGGER = getLogger(ChangelogService::class.java)
    }
}